src/common/autovector.h

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) 2018-Present Red Hat Inc.  All rights reserved.
+//
+// Copyright (c) 2011-2018, Facebook, Inc.  All rights reserved.
+// This source code is licensed under both the GPLv2 and Apache 2.0 License
+
+#ifndef CEPH_AUTOVECTOR_H
+#define CEPH_AUTOVECTOR_H
+
+#include <algorithm>
+#include <cassert>
+#include <initializer_list>
+#include <iterator>
+#include <stdexcept>
+#include <vector>
+
+// A vector that leverages pre-allocated stack-based array to achieve better
+// performance for array with small amount of items.
+//
+// The interface resembles that of vector, but with less features since we aim
+// to solve the problem that we have in hand, rather than implementing a
+// full-fledged generic container.
+//
+// Currently we don't support:
+//  * reserve()/shrink_to_fit()
+//     If used correctly, in most cases, people should not touch the
+//     underlying vector at all.
+//  * random insert()/erase(), please only use push_back()/pop_back().
+//  * No move/swap operations. Each autovector instance has a
+//     stack-allocated array and if we want support move/swap operations, we
+//     need to copy the arrays other than just swapping the pointers. In this
+//     case we'll just explicitly forbid these operations since they may
+//     lead users to make false assumption by thinking they are inexpensive
+//     operations.
+//
+// Naming style of public methods almost follows that of the STL's.
+namespace ceph {
+
+template <class T, size_t kSize = 8>
+class autovector {
```

Comment:
would [boost::container::small_vector](https://www.boost.org/doc/libs/1_67_0/doc/html/boost/container/small_vector.html) work for this?

Reply:
@bodley oh, I forgot about boost:container::small_vector.  Yes, I imagine it would!

Reply:
@markhpc i also think `boost::container::small_vector` might be a better choice here. could you reconsider this proposal?

Reply:
@tchaikov @cbodley Absolutely!  I've been a little hesitant to change it as part of this PR though as I already did a fair amount of testing using their existing autovector code and I'd like to make the change to small_vector as a separate future PR.  Does that seem reasonable?